### ISSUE TITLE
New metric generation and reporting

### ISSUE THREAD

**Issue by dbuscombe-usgs** [Doodleverse/segmentation_gym#40]
First, I think adopting a similar approach to [this](https://github.com/zhiminwang1/Remote-Sensing-Image-Segmentation/blob/master/seg_metrics.py), i.e. generating all stats from the CM, and also reporting the CM, makes much more sense

Also, explore Matthew's Correlation coefficient 

https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6413-7, generalized to multiclass

**Comment by dbuscombe-usgs** [comment]
- [ ] Implement what is discussed here #37

**Comment by dbuscombe-usgs** [comment]
I think this could be a utilities function. A stand-alone function that will take a validation subset, and model predictions on that subset, and create a confusion matrix, then report an array of stats from the confusion matrix:

- [ ] Overall Accuracy
- [ ] Precision
- [ ] Recall
- [ ] F1 Score
- [ ] Intersection Over Union 
- [ ] Mean Intersection Over Union
- [ ] Frequency Weighted Intersection over Union

as well as the following stats not generated (to my knowledge) by the c/m 

- [ ] Kullback Leibler divergence
- [ ] Matthews Correlation Coefficient

**Comment by dbuscombe-usgs** [comment]
that way, re-generating the stats would only require running `do_train` again using `do_train=False` in the config file

**Comment by dbuscombe-usgs** [comment]
Lots of good ideas here https://nowosad.github.io/post/motif-bp2/

**Comment by dbuscombe-usgs** [comment]
This is still important to look at. On multiclass problems, mean IOU is often pegged at 1.0 from the start of model training. This is not correct. Also, multiple metrics would be useful for some downstream model applications

**Comment by dbuscombe-usgs** [comment]
A new issue is that KLD appears always to be infinity for multiclass models. @CameronBodine are you seeing the same thing?

**Comment by CameronBodine** [comment]
Yes, KLD on train and validation subsets is always infinity.  I don't recall it being an issue when doing my binary depth model.

**Comment by dbuscombe-usgs** [comment]
KLD = infinity was caused by non-normalized, non-integer model outputs

solution is to compute the `argmax`, the one-hot encode

```
            kl = tf.keras.losses.KLDivergence() #initiate object

            est_label = np.argmax(est_label.squeeze(),axis=-1) #argmax to flatten()

            #one-hot encode
            nx,ny = est_label.shape
            lstack = np.zeros((nx,ny,NCLASSES))
            lstack[:,:,:NCLASSES+1] = (np.arange(NCLASSES) == 1+est_label[...,None]-1).astype(int) 

            #compute on one-hot encoded integer tensors
            kld = kl(tf.expand_dims(tf.squeeze(lbl), 0), lstack).numpy()
```

**Comment by dbuscombe-usgs** [comment]
I have implemented all of the metrics on [this](https://github.com/zhiminwang1/Remote-Sensing-Image-Segmentation/blob/master/seg_metrics.py) codebase, i.e. generating all stats from the CM

Need to update docs to credit the above

For example, this is a random model output (bad). The mean IoU much more accurately reflects the situation (generally bad), as does mean KLD, which is now fixed

```
Mean of mean IoUs (validation subset)=1.000
Mean of mean IoUs, confusion matrix (validation subset)=0.130
Mean of mean frequency weighted IoUs, confusion matrix (validation subset)=0.238
Mean of mean Dice scores (validation subset)=0.873
Mean of mean KLD scores (validation subset)=1.329
```

will appear in `train_model.py` as part of the final validation step on 10 batches of validation samples using a modified `plotcomp_n_metrics`

this function also creates two files of per-image model metrics. example files on a small number of validation samples below (just for illustration)

[noaa_spring2022_resunet_model1_model_metrics_per_sample.csv](https://github.com/Doodleverse/segmentation_gym/files/9466574/noaa_spring2022_resunet_model1_model_metrics_per_sample.csv)
[noaa_spring2022_resunet_model1_model_metrics_per_sample_per_class.csv](https://github.com/Doodleverse/segmentation_gym/files/9466575/noaa_spring2022_resunet_model1_model_metrics_per_sample_per_class.csv)

these mods require a mod to `doodleverse_utils.py` that will call all the model metrics in its own py script, useful also for Zoo and other extensibility

**Comment by dbuscombe-usgs** [comment]
started a `new  metrics` [branch](https://github.com/Doodleverse/segmentation_gym/tree/new_metrics) with these changes implemented. One could use this branch to use `train_model.py` to generate new sets of metrics for already trained models, usign `do_train`:false in the config file

Next is keras implementation iof the [matthews correlation coefficient](https://www.tensorflow.org/addons/api_docs/python/tfa/metrics/MatthewsCorrelationCoefficient). Looks straightforward:

```
>>> metric = tfa.metrics.MatthewsCorrelationCoefficient(num_classes=2)
>>> metric.update_state(y_true, y_pred)
>>> result = metric.result()
>>> result.numpy()
-0.33333334
```

Requires `pip install tensorflow_addons` in the conda env

**Comment by dbuscombe-usgs** [comment]
New `doodleverse_utils` version 0.0.4 is posted that contains new model metrics. For now, this is only required for users of the `new  metrics` [branch](https://github.com/Doodleverse/segmentation_gym/tree/new_metrics) of segmentation gym. 

`pip install doodleverse-utils -U` to upgrade from an existing activated gym repository

further, I have now tested the code using an already-trained model, for multiclass, and binary problems

here is a plot of metrics for a binary (water/no water) model

![tmp](https://user-images.githubusercontent.com/3596509/188032943-e38d7d5f-1fd5-41d5-8ce4-7cb78a003cd9.png)

**Comment by dbuscombe-usgs** [comment]
I believe I can close this issue, but first will do some tests with the new functions. In due course, @CameronBodine it would be helpful if you could checkout the `new_metrics` branch, update `doodleverse_utils`, and trial on your greyscale multiclass model - ta!
